SimTracker/TrackTriggerAssociation/interface/TTTrackAssociationMap.h

```diff
@@ -39,46 +51,57 @@ class TTTrackAssociationMap {
+
+  /// Set stub <-> truth association object.
+  void setTTStubAssociationMap(edm::RefProd<TTStubAssociationMap<T>> aStubAssoMap) {
     theStubAssociationMap = aStubAssoMap;
   }
 
-  /// Operations
-  edm::Ptr<TrackingParticle> findTrackingParticlePtr(edm::Ptr<TTTrack<T> > aTrack) const;
-  std::vector<edm::Ptr<TTTrack<T> > > findTTTrackPtrs(edm::Ptr<TrackingParticle> aTrackingParticle) const;
-
-  /// MC Truth methods
-  bool isGenuine(edm::Ptr<TTTrack<T> > aTrack) const;
-  bool isLooselyGenuine(edm::Ptr<TTTrack<T> > aTrack) const;
-  bool isCombinatoric(edm::Ptr<TTTrack<T> > aTrack) const;
-  bool isUnknown(edm::Ptr<TTTrack<T> > aTrack) const;
-
+  /// Get principle TP associated to a L1 track. (Non-NULL if isLooselyGenuine() below is true).
+  /// (N.B. There is no function returning all TP associated to a L1 track).
+  TrackingParticlePtr findTrackingParticlePtr(TTTrackPtrT<T> aTrack) const;
+
+  /// Get all L1 tracks associated to a TP.
+  /// (Even if the TP just contributes to one cluster in one stub,
+  /// and even if their are other such TP, it is still listed here).
+  std::vector<TTTrackPtrT<T>> findTTTrackPtrs(TrackingParticlePtr aTrackingParticle) const;
+
+  ///--- Get quality of L1 track based on truth info.
+  /// (N.B. "genuine" tracks are used for official L1 track efficiency measurements).
+
+  /// Exactly one (i.e. not 0 or >= 2) unique TP contributes to every stub on the track.
+  /// (even if it is not the principle TP in a stub, or contributes to only one cluster
+  /// in the stub, it still counts).
+  /// N.B. If cfg param getAllowOneFalse2SStub() is true, then one incorrect stub in
+  /// a 2S module is alowed
+  /// ISSUE: a track with 4 stubs can be accepted if only 3 of its stubs are correct!
+  /// ISSUE: isLooselyGenuine() must also be true. So if 2 TPs match track, one with
+  /// an incorrect PS stub, both isGenuine() & isLooselyGenuine() will be false!
+  bool isGenuine(TTTrackPtrT<T> aTrack) const;
+  /// Same criteria as for "genuine" track, except that one incorrect stub in either
+  /// PS or 2S module is allowed, irrespective of value of cfg param getAllowOneFalse2SStub().
+  bool isLooselyGenuine(TTTrackPtrT<T> aTrack) const;
+  /// More than one stub on track is "unknown".
+  bool isUnknown(TTTrackPtrT<T> aTrack) const;
+  /// Both isLooselyGenuine() & isUnknown() are false.
+  bool isCombinatoric(TTTrackPtrT<T> aTrack) const;
+
+  // Cfg param allowing one incorrect 2S stub in "genuine" tracks.
   void setAllowOneFalse2SStub(bool allowFalse2SStub);
   bool getAllowOneFalse2SStub();
 
 private:
   /// Data members
-  std::map<edm::Ptr<TTTrack<T> >, edm::Ptr<TrackingParticle> > trackToTrackingParticleMap;
-  std::map<edm::Ptr<TrackingParticle>, std::vector<edm::Ptr<TTTrack<T> > > > trackingParticleToTrackVectorMap;
-  edm::RefProd<TTStubAssociationMap<T> > theStubAssociationMap;
+  MapL1TrackToTP<T> trackToTrackingParticleMap;
```

Comment:
member variable names should end in underscore

Reply:
done